apps/uv-test/app.c

```diff
@@ -0,0 +1,414 @@
+  SI1133_ADCMUX_MEDIUM_IR = 0x01,
+  SI1133_ADCMUX_LARGE_IR =  0x02,
+  SI1133_ADCMUX_WHITE =  0x0b,
+  SI1133_ADCMUX_LARGE_WHITE = 0x0d,
+  SI1133_ADCMUX_UV = 0x18,
+  SI1133_ADCMUX_UV_DEEP = 0x19,
+} si1133_adcmux;
+
+typedef enum {
+    SI1133_DECIM_0 = 0,
+    SI1133_DECIM_1 = 1,
+    SI1133_DECIM_2 = 2,
+    SI1133_DECIM_3 = 3,
+} si1133_decim;
+
+typedef enum {
+    SI1133_HW_GAIN_0 = 0,
+    SI1133_HW_GAIN_1 = 1,
+    SI1133_HW_GAIN_2 = 2,
+    SI1133_HW_GAIN_3 = 3,
+    SI1133_HW_GAIN_4 = 4,
+    SI1133_HW_GAIN_5 = 5,
+    SI1133_HW_GAIN_6 = 6,
+    SI1133_HW_GAIN_7 = 7,
+    SI1133_HW_GAIN_8 = 8,
+    SI1133_HW_GAIN_9 = 9,
+    SI1133_HW_GAIN_10 = 10,
+    SI1133_HW_GAIN_11 = 11,
+} si1133_hw_gain;
+
+typedef enum {
+    SI1133_RANGING_OFF = 0,
+    SI1133_RANGING_ON = 1
+} si1133_ranging;
+
+typedef enum {
+    SI1133_SW_GAIN_0 = 0,
+    SI1133_SW_GAIN_1 = 1,
+    SI1133_SW_GAIN_2 = 2,
+    SI1133_SW_GAIN_3 = 3,
+    SI1133_SW_GAIN_4 = 4,
+    SI1133_SW_GAIN_5 = 5,
+    SI1133_SW_GAIN_6 = 6,
+    SI1133_SW_GAIN_7 = 7,
+} si1133_sw_gain;
+
+typedef enum {
+    SI1133_16_BIT = 0,
+    SI1133_24_BIT = 1
+} si1133_measurement_size;
+
+typedef enum {
+    SI1133_POST_SHIFT_0 = 0,
+    SI1133_POST_SHIFT_1 = 1,
+    SI1133_POST_SHIFT_2 = 2,
+    SI1133_POST_SHIFT_3 = 3,
+} si1133_post_shift;
+
+typedef enum {
+    SI1133_THRESH_DISABLE = 0,
+    SI1133_THRESH_0 = 1,
+    SI1133_THRESH_1 = 2,
+    SI1133_THRESH_2 = 3,
+} si1133_threshold_config;
+
+typedef enum {
+    SI1133_COUNTER_DISABLE = 0,
+    SI1133_MEASCOUNT_1 = 1,
+    SI1133_MEASCOUNT_2 = 2,
+    SI1133_MEASCOUNT_3 = 3,
+} si1133_count_config;
+
+void error(char *msg);
+void si1133_init(void);
+
+/**
+ * @brief configure channel
+ */
+int si1133_configure_channel(
+        si1133_channels channel,
+        si1133_adcmux adcmux,
+        si1133_decim decim_rate,
+        si1133_ranging ranging,
+        si1133_hw_gain hw_gain,
+        si1133_sw_gain sw_gain,
+        si1133_measurement_size measurement_size,
+        si1133_post_shift post_shift,
+        si1133_threshold_config thresh_config,
+        si1133_count_config count_config
+        );
+
+/**
+ * @brief set value for given param in si1133 internal parameter table
+ * @param param address of parameter you want to set (6 bits long)
+ * @param value value you want set for given parameter
+ * @details
+ * See section 5.1 of the si1133 datasheet for more details
+ */
+int si1133_set_param(si1133_params param, uint8_t value);
+
+void app_init(void) {
+    watch_enable_display();
+
+    watch_enable_external_interrupts();
+    watch_register_interrupt_callback(BTN_MODE, cb_mode_pressed, INTERRUPT_TRIGGER_RISING);
+    watch_register_interrupt_callback(BTN_LIGHT, cb_light_pressed, INTERRUPT_TRIGGER_RISING);
+    watch_register_interrupt_callback(BTN_ALARM, cb_alarm_pressed, INTERRUPT_TRIGGER_RISING);
+
+    watch_enable_leds();
+
+    watch_register_interrupt_callback(A2, cb_si1133_int, INTERRUPT_TRIGGER_FALLING);
```

Comment:
@joeycastillo can you spot a mistake here somewhere? I can't get the interrupt to fire, even though the pin does go low

Reply:
This is a subtle but annoying issue with the ASF external interrupt framework; there's [a warning about it in the docs](https://joeycastillo.github.io/Sensor-Watch/group__buttons.html#gae40d2f73d0412680d83c26b8a0e1d31c). The gist is that pin A2 uses the same interrupt channel as the Alarm button, and the ASF framework only wants one pin associated with each channel. You could change that association [deep in the EIC config file](https://github.com/joeycastillo/Sensor-Watch/blob/main/watch-library/shared/config/hpl_eic_config.h#L726), but it would mean the Alarm button no longer works with the EIC.

My suggestion: instead of using `watch_register_interrupt_callback`, use `watch_register_extwake_callback` ([documented here](https://joeycastillo.github.io/Sensor-Watch/group__deepsleep.html#gacad726423f5f38697c87474ab25244ca)). This has the added benefit that the interrupt on A2 will be able to wake the watch from deep sleep.

Reply:
I guess revision 2 of the UV sensor board should use a different pin :sweat_smile: , thanks for the clarification, I didn't think to read the docs...

Reply:
TBH I think pins A2 and A4 are the best option for an external interrupt, since they can wake the watch from deep sleep (the EIC doesn't run in that deep sleep, but the external wake on those two pins does). Unless you need to trigger on both rising and falling edges, the extwake callback on A2 should do everything the interrupt callback would do for you. 

Having said that if you do want to use the EIC for whatever reason, I'd recommend pin A4 since it also gives you the option of waking from deep sleep on an interrupt signal.